### PR TITLE
[#111] Make styleId an optional string or function

### LINKED ISSUE #111: Make styleId an optional string or function

### PR DESCRIPTION
**Pull request by pedrosousa13**
We have a build that builds out css with `relativeCSSInjection`. I would like to have style tags with different id's as so not to clash.

Using this with the SSR approach that I modified in my style injection function works really well https://github.com/egoist/style-inject/pull/28/files

Styles are loaded via SSR with different `id` attributes. This will only run if `styleId` is specified.

I'm adding this PR because I need the id to be the same for server + client so I need it at the vite level.

closes #111

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -217,6 +228,7 @@ export async function relativeCssInjection(
 
         // We have already filtered these chunks to be RenderedChunks
         const jsAsset = bundle[jsAssetName] as OutputChunk;
+
```
Comment (marco-prontera) [inline src/utils.ts:231]: Remove empty line

File: src/utils.ts [hunk 2]
```diff
@@ -294,11 +306,10 @@ export function buildOutputChunkWithCssInjectionCode(
...
     const appCode = jsAssetCode.replace(/\/\*.*empty css.*\*\//, '');
-    jsAssetCode = topExecutionPriorityFlag ? '' : appCode;
-    jsAssetCode += cssInjectionCode;
-    jsAssetCode += !topExecutionPriorityFlag ? '' : appCode;
 
-    return jsAssetCode;
+    return topExecutionPriorityFlag
+      ? `${cssInjectionCode}${appCode}` :
+      `${appCode}${cssInjectionCode}`;
 }
```
Comment (marco-prontera) [inline src/utils.ts:313]: Leave the code as before please

File: src/interface.ts [hunk 3]
```diff
@@ -16,6 +17,7 @@ export interface PluginConfiguration extends BaseOptions {
     preRenderCSSCode?: (cssCode: string) => string;
     relativeCSSInjection?: boolean;
     suppressUnusedCssWarning?: boolean;
+    autoStyleId?: boolean;
```
Comment (marco-prontera) [inline src/interface.ts:20]: Unused remove please

File: src/interface.ts [hunk 4]
```diff
@@ -8,6 +8,7 @@ export interface BaseOptions {
     styleId?: string;
     topExecutionPriority?: boolean;
     useStrictCSP?: boolean;
+    relativeCSSInjection?: boolean;
```
Comment (marco-prontera) [inline src/interface.ts:11]: This should be moved to the interface that use is so BuildCSS...

File: src/utils.ts [hunk 5]
```diff
@@ -26,14 +27,24 @@ export async function buildCSSInjectionCode({
     injectCodeFunction,
     useStrictCSP,
     buildOptions,
+    relativeCSSInjection,
 }: BuildCSSInjectionConfiguration): Promise<OutputChunk | null> {
     let { minify, target } = buildOptions;
 
+    // Ensures styleId is unique when relativeCSSInjection is enabled
+    const generatedStyleId = styleId && relativeCSSInjection ? `${styleId ? `${styleId}-` : ''}${uuidv4()}` : styleId ;
```
Comment (marco-prontera) [inline src/utils.ts:35]: Check the length of styleId maybe can be an empty string? I can't check this right now, maybe later

**Review by marco-prontera — CHANGES_REQUESTED** [review]
Thank you for contribution!
After changes, please write some test for these changes

File: src/utils.ts (hunk 2 above)
Comment (pedrosousa13) [reply src/utils.ts:313]: sure, I can leave it as is. is my implementation wrong?

**Comment by pedrosousa13** [comment]
@marco-prontera I made the requested changes.

**Review by marco-prontera — CHANGES_REQUESTED** [review]
Hi @pedrosousa13, I talked to other contributors about what they think and we ended up not introducing this dependency but changing the styleId so that it is also possible to configure a callback which in this case will be called every time for every css file injection. 
Can you change the code?

**Comment by pedrosousa13** [comment]
yeah totally. will make the change.

**Comment by pedrosousa13** [comment]
@marco-prontera changes made, it now accepts a callback function. I also like this implementation :)

**Review by marco-prontera — APPROVED** [review]

**Comment by marco-prontera** [comment]
Hi @pedrosousa13, I published the new version [3.3.0](https://www.npmjs.com/package/vite-plugin-css-injected-by-js)
Thank you for the contribution 🥳

**Pull request merged**